File: Hot 100/98. 验证二叉搜索树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root, long left=LONG_MIN,long right=LONG_MAX) {
        if (root==nullptr){
            return true;
        }
        long x=root->val;
        return left<x&&x<right && isValidBST(root->left,left,x) && isValidBST(root->right,x,right);
    }
};

File: Hot 100/98. 验证二叉搜索树.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return check(root, math.MinInt64, math.MaxInt64)
}
func check(root *TreeNode, low int, high int) bool {
	if root == nil {
		return true
	}
	if root.Val <= low || root.Val >= high {
		return false
	}
	return check(root.Left, low, root.Val) && check(root.Right, root.Val, high)
}